### PR TITLE
feat: added a CodePipeline Docker Python example

### PR DESCRIPTION
**Pull request by josjaf**
Addressing comments https://github.com/aws-samples/aws-cdk-examples/pull/138

### CODE REVIEW

File: python/codepipeline-docker-build/Pipeline.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+from aws_cdk import (
+
```
Comment (skinny85) [inline python/codepipeline-docker-build/Pipeline.py]: Unnecessary empty line.

File: python/codepipeline-docker-build/Pipeline.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        pipeline = aws_codepipeline.Pipeline(
+            self, "Pipeline",
+            pipeline_name=f"{props.namespace}",
+            artifact_bucket=props.bucket_obj,
+            stages=[
+                aws_codepipeline.StageProps(
+                    stage_name='Source',
+                    actions=[
+                        aws_codepipeline_actions.S3SourceAction(
+
```
Comment (skinny85) [inline python/codepipeline-docker-build/Pipeline.py]: Unnecessary empty line.

File: python/codepipeline-docker-build/Pipeline.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                            bucket=props.bucket_obj,
+                            bucket_key='source.zip',
+                            action_name='S3Source',
+                            run_order=1,
+                            output=source_output,
+
```
Comment (skinny85) [inline python/codepipeline-docker-build/Pipeline.py]: Unnecessary empty line.

File: python/codepipeline-docker-build/Pipeline.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                        ),
+                    ]
+                ),
+                aws_codepipeline.StageProps(
+
```
Comment (skinny85) [inline python/codepipeline-docker-build/Pipeline.py]: Unnecessary empty line.

File: python/codepipeline-docker-build/Pipeline.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                    stage_name='Build',
+                    actions=[aws_codepipeline_actions.CodeBuildAction(
+                        action_name='DockerBuildImages',
+                        # role=codepipeline_role,
+                        input=source_output,
+                        project=props.cb_docker_build,
+                        run_order=1,
+
```
Comment (skinny85) [inline python/codepipeline-docker-build/Pipeline.py]: Unnecessary empty line.

File: python/codepipeline-docker-build/Pipeline.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+                            run_order=1,
+                            output=source_output,
+
+                        ),
+                    ]
+                ),
+                aws_codepipeline.StageProps(
+
+                    stage_name='Build',
+                    actions=[aws_codepipeline_actions.CodeBuildAction(
```
Comment (skinny85) [inline python/codepipeline-docker-build/Pipeline.py]: Indent is weird here - should be

```python
                      actions=[
                        aws_codepipeline_actions.CodeBuildAction(
                          action_name='DockerBuildImages',
                          # role=codepipeline_role,
                          input=source_output,
                          project=props.cb_docker_build,
                          run_order=1,
                        ),
                      ],
```

like it is above.

File: python/codepipeline-docker-build/Pipeline.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+                        action_name='DockerBuildImages',
+                        # role=codepipeline_role,
```
Comment (skinny85) [inline python/codepipeline-docker-build/Pipeline.py]: I guess this line can be removed? It's probably not a great idea to have commented out code in an example :)

File: python/codepipeline-docker-build/app.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+from Base import Base
+from Pipeline import Pipeline
+
+
+# using props to pass in objects between stacks
+class Props():
+    def __init__(self):
+        self.namespace = 'cdk-example-pipeline'
+        self.region = 'us-east-1'
```
Comment (skinny85) [inline python/codepipeline-docker-build/app.py]: I don't see this parameter being read anywhere - is it actually used?

File: python/codepipeline-docker-build/ReadMe.md [hunk 9]
```diff
@@ -0,0 +1,6 @@
+# CDK Python Codepipeline Example
```
Comment (skinny85) [inline python/codepipeline-docker-build/ReadMe.md]: Nitpick - missing empty line after`# CDK Python CodePipeline (capital "P") example`.

File: python/codepipeline-docker-build/ReadMe.md [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+* This is an example of a CodePipeline project that uses CodeBuild to Build a Docker Image and push to ECR.
+* This example uses multiple stacks for the purpose of demonstrating ways of passing in objects from different stacks
```
Comment (skinny85) [inline python/codepipeline-docker-build/ReadMe.md]: Nitpick - missing period at the end of the sentence.

File: python/codepipeline-docker-build/Base.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+        props.bucket_name = bucket.bucket_name
+        props.bucket_arn = bucket.bucket_arn
+        props.bucket_obj = bucket
+        props.cb_docker_build = cb_docker_build
+        self.output_props = props
+
+    # pass objects to another stack
+    @property
+    def outputs(self):
+        props = self.output_props
```
Comment (skinny85) [inline python/codepipeline-docker-build/Base.py]: I think `return self.output_props` is all you need here :)

File: python/codepipeline-docker-build/app.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+
+props = Props()
+app = core.App()
+
+# stack for ecr, bucket, codebuild
+base = Base(app, f"{props.namespace}-base", props, )
```
Comment (skinny85) [inline python/codepipeline-docker-build/app.py]: No need for the comma after `props` (the comma is only needed for multi-line calls).

File: python/codepipeline-docker-build/Base.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+        # codebuild permissions to interact with ecr
+        ecr.grant_pull_push(cb_docker_build)
+
+        # update props to pass objects to another stack
+        props.bucket_name = bucket.bucket_name
+        props.bucket_arn = bucket.bucket_arn
+        props.bucket_obj = bucket
+        props.cb_docker_build = cb_docker_build
+        self.output_props = props
```
Comment (skinny85) [inline python/codepipeline-docker-build/Base.py]: I would skip modifying `props`, I feel like it's more complicated than necessary. Just save the updated props locally.

File: python/codepipeline-docker-build/app.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+props = base.outputs
+
+# pipeline stack
+pipeline = Pipeline(app, f"{props.namespace}-pipeline", props)
```
Comment (skinny85) [inline python/codepipeline-docker-build/app.py]: I would skip modifying `props`. Having the same variable mean different things in the program is just more overhead to me. Just pass `base.outputs` here.

File: python/codepipeline-docker-build/push.sh [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+
+
+export account_id=$(aws sts get-caller-identity | jq -r .Account)
+export source_bucket=$(aws ssm get-parameter --name 'cdk-example-pipeline-bucket' | jq -r .Parameter.Value)
+export pipeline_name=$(aws ssm get-parameter --name 'cdk-example-pipeline-pipeline' | jq -r .Parameter.Value)
+export REGION='us-east-1'
+
+zip -r source.zip .
+aws s3 cp source.zip s3://${source_bucket}/source.zip
+aws codepipeline start-pipeline-execution --name ${pipeline_name}
```
Comment (skinny85) [inline python/codepipeline-docker-build/push.sh]: You don't need this, right? Just pushing to the source bucket should trigger the pipeline, right?

File: python/codepipeline-docker-build/requirements.txt [hunk 16]
```diff
@@ -0,0 +1,49 @@
+attrs==19.2.0
+aws-cdk.assets==1.12.0
```
Comment (skinny85) [inline python/codepipeline-docker-build/requirements.txt]: We don't hard-code the version of the CDK libraries  - just leave it as `aws-cdk.assets`. Same for the other dependencies.

File: python/codepipeline-docker-build/push.sh (hunk 15 above)
Comment (josjaf) [reply python/codepipeline-docker-build/push.sh]: This way is faster triggering than polling for changes. Guess it doesn't really matter one way or the other.

File: python/codepipeline-docker-build/app.py (hunk 14 above)
Comment (josjaf) [reply python/codepipeline-docker-build/app.py]: I like the idea of passing around an object that is independent of a single stack. If we had a third stack here that needed objects from stack 1 and 2, I think this way works better - unless you were thinking of something else?

File: python/codepipeline-docker-build/Base.py (hunk 13 above)
Comment (josjaf) [reply python/codepipeline-docker-build/Base.py]: I'm not following, what do you mean?

File: python/codepipeline-docker-build/Pipeline.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+                            input=source_output,
+                            project=props['cb_docker_build'],
+                            run_order=1,
+                        )
+                    ]
+                )
+            ]
+
+        )
+        print(props['bucket'])
```
Comment (skinny85) [inline python/codepipeline-docker-build/Pipeline.py]: This should probably be gone :)

**Comment by josjaf** [comment]
I just fixed something of the things we talked about and pushed

**Pull request merged**